File: SVD/Python/randomizedSVD.py
import numpy as np

def randomizedSVD(A,l):
    m,n = A.shape
    Omega = np.random.randn(n,l)
    Y = A @ Omega
    Q, _ = np.linalg.qr(Y, mode='complete')
    B = Q @ A
    Sigma=np.zeros((m,n))
    U_tilde, S, V = np.linalg.svd(B, full_matrices=True, compute_uv=True) #viene già restituita V trasposta
    U = Q @ U_tilde
    Sigma[:m,:m] = np.diag(S)
    return U, Sigma, V

def randomizedSVD_v2(A,l,q):
    m, n = A.shape
    Omega = np.random.randn(n, l)
    Y0 = A @ Omega
    Q0, R0 = np.linalg.qr(Y0, mode='complete')
    
    for i in range(1, q):
        Yi = A.T @ Q0
        Qi, Ri = np.linalg.qr(Yi, mode='complete')
        Y0 = A @ Qi
        Q0, R0 = np.linalg.qr(Y0, mode='complete')
    
    B = Q0 @ A  
    Sigma = np.zeros((m, n))
    U_tilde, S, V = np.linalg.svd(B, full_matrices=True, compute_uv=True)
    U = Q0 @ U_tilde
    Sigma[:m, :m] = np.diag(S) 
    
    return U, Sigma, V


def randomizedSVD_v3(A,l,q):
   m,n = A.shape
   Omega = np.random.randn(n,l)
   prodotto=A.T@A
   for i in range(1,q):
    prodotto=prodotto@prodotto
   Y=A @ Omega
   Q,_ = np.linalg.qr(Y, mode='complete')
   B = Q @ A
   Sigma=np.zeros((m,n))
   U_tilde, S, V = np.linalg.svd(B, full_matrices=True, compute_uv=True)
   U = Q @ U_tilde
   Sigma[:m,:m] = np.diag(S)
   return U, Sigma, V




A=np.random.randn(100,200)
l=5
q=5
U, S, V = randomizedSVD(A,l)
A_reconstructed = U @ S @ V
print(np.linalg.norm(A-A_reconstructed,ord='fro'))
U, S, V = randomizedSVD_v2(A,l,q)
A_reconstructed = U @ S @ V
print(np.linalg.norm(A-A_reconstructed,ord='fro'))
U, S, V = randomizedSVD_v3(A,l,q)
A_reconstructed = U @ S @ V
print(np.linalg.norm(A-A_reconstructed,ord='fro'))

File: FontanaEsame/Implementazioni/SVD/MatLab/randomizedSVD.m
A = randn(10, 20);
l=5;
q=5;
[U,S,V] = randomizedSVD_v1(A,l);
A_ricostruita = U*S*V';
disp("NORMA DIFFERENZA = "+norm(A-A_ricostruita));
[U,S,V] = randomizedSVD_v2(A,l,q);
A_ricostruita = U*S*V';
disp("NORMA DIFFERENZA = "+norm(A-A_ricostruita));
[U,S,V] = randomizedSVD_v3(A,l,q);
A_ricostruita = U*S*V';
disp("NORMA DIFFERENZA = "+norm(A-A_ricostruita));



function [U, S, V] = randomizedSVD_v1(A, l)
    [m, n] = size(A);
    Omega = randn(n, l);
    Y = A * Omega;
    [Q, ~] = qr(Y);
    B = Q * A;
    [U_tilde, S, V] = svd(B);
    U = Q * U_tilde;
end



function [U,S,V] = randomizedSVD_v2(A,l,q)
    [m,n] = size(A);
    Omega = randn(n,l);
    Y0 = A*Omega;
    [Q0,~] = qr(Y0);
    for i = 1:q
        Yj = A' * Q0;
        [Qj,~] = qr(Yj);
        Y0 = A * Qj;
        [Q0,~] = qr(Y0);
    end
    B = Q0*A;
    [U_tilde,S,V] = svd(B);
    U = Q0*U_tilde;
end

function [U,S,V] = randomizedSVD_v3(A,l,q)
    [m,n] = size(A);
    Omega = randn(m,l);
    Y=A*A';
    for i = 1:q
        Y = Y*Y;
    end
    [Q, ~] = qr(Y);
    B = Q * A;
    [U_tilde, S, V] = svd(B);
    U = Q * U_tilde;
end



